### PR TITLE
Do not use the Lsp-Misc workspace in VS scenarios

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Tracking down why we're seeing some flakeyness in LSP pull diagnostic scenarios.  The misc workspace is showing up unepectedly, so this tries to beef things up to make things better.

This is slightly more aggressive than https://github.com/dotnet/roslyn/pull/67289 in terms of teh change.

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceRegistrationEventListener.cs [hunk 1]
```diff
@@ -9,8 +9,16 @@
...
-[ExportEventListener(WellKnownEventListeners.Workspace, WorkspaceKind.Host, WorkspaceKind.MiscellaneousFiles, WorkspaceKind.MetadataAsSource, WorkspaceKind.MSBuild, WorkspaceKind.Interactive), Shared]
+[ExportEventListener(
+    WellKnownEventListeners.Workspace,
+    WorkspaceKind.Host,
+    WorkspaceKind.MiscellaneousFiles,
+    WorkspaceKind.MetadataAsSource,
+    // TODO(cyrusn): Why does LSP need to know about the msbuild workspace?  It's a workspace that is used in console
+    // apps, not rich server scenarios.
+    WorkspaceKind.MSBuild,
+    // TODO(cyrusn): Why does LSP need to know about the interactive workspace? Does LSP work in interactive buffers?
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceRegistrationEventListener.cs:20]: It does , for example when the all lsp flag is on, it powers completion in interactive.

File: src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceRegistrationEventListener.cs [hunk 2]
```diff
@@ -9,8 +9,16 @@
...
 namespace Microsoft.CodeAnalysis.LanguageServer;
 
-// Specify the VS workspaces we know we need (host, misc, metadata as source) and MSBuild for VSCode.
-[ExportEventListener(WellKnownEventListeners.Workspace, WorkspaceKind.Host, WorkspaceKind.MiscellaneousFiles, WorkspaceKind.MetadataAsSource, WorkspaceKind.MSBuild, WorkspaceKind.Interactive), Shared]
+[ExportEventListener(
+    WellKnownEventListeners.Workspace,
+    WorkspaceKind.Host,
+    WorkspaceKind.MiscellaneousFiles,
+    WorkspaceKind.MetadataAsSource,
+    // TODO(cyrusn): Why does LSP need to know about the msbuild workspace?  It's a workspace that is used in console
```
Comment (dibarbet) [inline src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceRegistrationEventListener.cs:17]: probably not necessary anymore

**Review by dibarbet — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceRegistrationEventListener.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceRegistrationEventListener.cs:20]: fascinating.

File: src/Features/LanguageServer/Protocol/RoslynLanguageServer.cs [hunk 3]
```diff
@@ -82,6 +81,13 @@ protected override IRequestExecutionQueue<RequestContext> ConstructRequestExecut
             AddBaseService<IMethodHandler>(new InitializedHandler());
             AddBaseService<IOnInitialized>(this);
 
+            // In all VS cases, we already have a misc workspace.  Specifically
+            // Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.MiscellaneousFilesWorkspace.  In
+            // those cases, we do not need to add an additional workspace to manage new files we hear about.  So only
+            // add the LspMiscellaneousFilesWorkspace for hosts that have not already brought their own.
+            if (serverKind == WellKnownLspServerKinds.CSharpVisualBasicLspServer)
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/RoslynLanguageServer.cs:88]: slightly worried about brittleness here.  @dibarbet is this ok for WellKnownLspServerKinds.Razor?

Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/RoslynLanguageServer.cs:88]: I believe it should be fine - the `CSharpVisualBasicLspServer` should be the only server that is used outside of VS.

**Review by dibarbet — APPROVED** [review]

**Pull request merged**